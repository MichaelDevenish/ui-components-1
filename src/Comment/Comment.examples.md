## A Sample Comment Component

```jsx
<Comment
  comment='This is a test comment'
/>
```
