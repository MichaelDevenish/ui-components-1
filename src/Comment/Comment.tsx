import React from 'react'

export interface CommentProps {
  comment: string
}

export const Comment: React.SFC<CommentProps> = props => (
  <div>
    {props.comment}
  </div>
)