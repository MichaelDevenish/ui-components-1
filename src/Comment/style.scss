:local(.Comment) {
  display: flex;
  flex-flow: row;

  >.comment-badge-container {
    align-self: flex-start;
    padding: .5rem;
  }

  >.comment-content-container {
    align-self: stretch;
    flex-grow: 1;
    padding: .6rem 1.2rem;
    border-radius: .75rem;
    background-color: $neutral-light;

    >.comment-header-container {
      display: flex;
      flex-flow: row;
      align-items: center;
      justify-content: center;

      >.comment-header-person-name {
        flex-grow: 1;
      }

      >.comment-header-owner-label {
        font-size: 12px;
        border-radius: 31px;
        padding: 4px 11px;
        font-weight: 400;
        border: 1px solid $neutral-base;
      }

      >.comment-header-date {
        font-size: 80%;
        margin-left: .75rem;
      }
    }
  }
}