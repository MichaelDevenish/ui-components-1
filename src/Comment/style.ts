import React from 'react'
import styled, { StyledComponentClass } from 'styled-components'
import { getColor } from '../Color'

const ActionMenuButtonToggleButton = styled.button`
  background-color: transparent;
  color: ${getColor('badge-background-dark')};
  cursor: pointer;
  padding: .2rem;
  transition: background-color .25s ease-out, color .25s ease-out;

  &:hover,
  &:active,
  &:focus {
    background-color: ${getColor('neutral-base')};
    color: ${getColor('main-text')};
    outline: none;
  }

  .fa {
    font-size: 1rem;
    line-height: 1;
    margin: 0;
    width: 1.3em;
  }
`

export {
  ActionMenuButtonToggleButton
}
