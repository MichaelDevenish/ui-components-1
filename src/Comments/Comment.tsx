import React from 'react'
import classNames from 'classnames'
import { PersonBadge } from '../PersonBadge/PersonBadge'
const style = require('./style.scss')

export interface CommentProps {
  comment: {
    id: string,
    comment: string,
    header?: React.Component,
    personDisplayName: string,
    personFirstName: string,
    personPreferredOrFirstName: string,
    createdAt: string,
    personId: string,
    personProfilePictureId?: string,
    label?: React.Component
  }
  loggedInUser: {
    id: string
  }
  idx: number
  alternatingColours?: boolean
}

export class Comment extends React.Component<CommentProps> {
  get personProfilePicture (): JSX.Element {
    const {
      personFirstName,
      personPreferredOrFirstName,
      personProfilePictureId
    } = this.props.comment

    return (
      <div className='comment-badge-container'>
        <PersonBadge
          size='small'
          preferredOrFirstName={personPreferredOrFirstName}
          lastName={personFirstName}
        />
      </div>
    )
  }

  get ownerLabel (): JSX.Element | null {
    const {
      comment: {
        personId: commenterId
      },
      loggedInUser: {
        id: loggedInUserId
      }
    } = this.props

    if (commenterId === loggedInUserId) {
      return <div className='comment-header-owner-label'>Owner</div>
    }

    return null
  }

  get commentDate (): JSX.Element {
    return (
      <div className='comment-header-date'>
        25 minutes ago
      </div>
    )
  }

  get commenterName (): JSX.Element {
    const {
      comment: {
        personId,
        personDisplayName
      },
      loggedInUser: {
        id: loggedInUserId
      }
    } = this.props

    console.log(personId, loggedInUserId)

    return (
      <strong className='comment-header-person-name'>
        {personId === loggedInUserId ? 'You' : personDisplayName}
      </strong>
    )
  }

  get commentHeaderMeta (): React.Component | null {
    const {
      comment: {
        header
      }
    } = this.props

    if (header) {
      return header
    }

    return null
  }

  get commentHeaderLabel (): React.Component | null {
    const {
      comment: {
        label
      }
    } = this.props

    if (label) {
      return label
    }

    return null
  }

  get commentHeader (): JSX.Element {
    return (
      <div className='comment-header-container'>
        {this.commenterName}
        {this.commentHeaderMeta}

        {this.ownerLabel}
        
        {this.commentDate}
        {this.commentHeaderLabel}
      </div>
    )
  }

  get commentContent (): JSX.Element {
    const {
      comment
    } = this.props.comment

    return (
      <div className='comment-content'>
        {comment}
      </div>
    )
  }

  render (): JSX.Element {
    const {
      comment
      // idx,
      // alternatingColours
    } = this.props

    return (
      <div id={comment.id} className={classNames(style.Comment)}>
        {this.personProfilePicture}
        <div className='comment-content-container'>
          {this.commentHeader}
          {this.commentContent}
        </div>
      </div>
    )
  }
}