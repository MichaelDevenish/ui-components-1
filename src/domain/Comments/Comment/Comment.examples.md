#### Someone else's comment
```jsx
<Comment
  comment={{
    id: 'sampleCommentId',
    comment: 'A very recent comment',
    personDisplayName: 'Example Person Name',
    personLastName: 'Culkin',
    personPreferredOrFirstName: 'McKooley',
    createdDateText: <span>just now</span>,
    personId: 'sampleCommentPersonId'
  }}
  loggedInUser={{
    id: 'sampleLoggedUserId'
  }}
/>
```

#### Own comment
```jsx
<Comment
  comment={{
    id: 'sampleCommentId',
    comment: 'This is some ancient comment',
    personDisplayName: 'Example Person Name',
    personLastName: 'Culkin',
    personPreferredOrFirstName: 'McKooley',
    createdDateText: <span>1st Jan 2017</span>,
    personProfilePictureId: 'http://www.multiplemayhemmamma.com/wp-content/uploads/2013/03/home-alone-150x150.jpg',
    personId: 'sampleCommentPersonId'
  }}
  loggedInUser={{
    id: 'sampleCommentPersonId'
  }}
/>
```

### Own comment with some actions attached to it
```jsx
<Comment
  comment={{
    id: 'sampleCommentId',
    comment: 'This is some ancient comment',
    personDisplayName: 'Example Person Name',
    personLastName: 'Culkin',
    personPreferredOrFirstName: 'McKooley',
    createdDateText: <span>1st Jan 2017</span>,
    personProfilePictureId: 'http://www.multiplemayhemmamma.com/wp-content/uploads/2013/03/home-alone-150x150.jpg',
    personId: 'sampleCommentPersonId'
  }}
  actions={[{
    text: 'Edit',
    onClick: () => { alert('Edit handler') }
  },{
    component: <hr />
  },{
    text: 'Delete',
    onClick: () => { alert('Delete handler') },
    sectionType: 'alert'
  }]}
  loggedInUser={{
    id: 'sampleCommentPersonId'
  }}
/>
```

### Comments that have a header defined will not render actionMenu and will be considered a system generated comment
```jsx
const { Text } = require('@Domain/Typographies');
const { Variables } = require('@Common');

<Comment
  comment={{
    id: 'sampleCommentId',
    personDisplayName: 'Lyanna Moreton',
    personLastName: 'Moreton',
    personPreferredOrFirstName: 'Lyanna',
    createdDateText: <span>1st Jan 2017</span>,
    personProfilePictureId: 'http://www.multiplemayhemmamma.com/wp-content/uploads/2013/03/home-alone-150x150.jpg',
    personId: 'lyanna',
    header: (
      <span style={{ marginLeft: '5px' }}>
        has
        <Text
          isHeavy
          color={Variables.Color.g400}
        >
          approved
        </Text>
        Joshua Brady's
        <Text
          isHeavy
          color={Variables.Color.i400}
        >
          qualification
        </Text>
        with a very long message to test how a long message would look.
      </span>
    )
  }}
  loggedInUser={{
    id: 'sampleCommentPersonId'
  }}
/>
```
