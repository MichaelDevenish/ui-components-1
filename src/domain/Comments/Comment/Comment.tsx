import classNames from 'classnames'
import React from 'react'

import { FontAwesomeIcon } from '../../Icons'
import {
  DropdownMenu,
  IDropdownMenuSectionProps,
  IDropdownMenuToggleComponentProps
} from '../../Popovers'
import { FormattedText } from '../../Typographies'
import {
  CommentActionMenuToggleButton,
  StyledComment
} from './style'

export interface ICommentProps {
  /** the name of the person who posted the comment */
  commentHeaderText: string,
  /** Displays the comment in a focused state */
  focused?: boolean,
  /** The component to render to the left of the comment box */
  avatarComponent: JSX.Element,
  /** A pill to be displayed right before the date label of the comment (use a span to persist the inline diplay behaviour) */
  pillComponent?: JSX.Element,
  /** Additional text to display in the array header (preferable a span HTML element) */
  headerComponent?: JSX.Element,
  /** Use DateText component from the list of alreay defined components (a string should work too) */
  dateComponent: JSX.Element,
  /** Comment text string */
  commentBodyText: string,
  /** an array of sectionProps from the DropdownMenu to render the submenu sections */
  actions?: IDropdownMenuSectionProps[]
}

export class Comment extends React.Component<ICommentProps> {
  public render (): JSX.Element {
    const {
      focused
    } = this.props

    return (
      <StyledComment>
        {this.avatar}
        <div
          className={classNames(
            'comment-container',
            { focused }
          )}
        >
          {this.commentHeader}
          {this.commentContent}
        </div>
      </StyledComment>
    )
  }

  private get avatar (): JSX.Element {
    const {
      avatarComponent
    } = this.props

    return (
      <div className='comment-badge-container'>
        {avatarComponent}
      </div>
    )
  }

  private get commentActions (): JSX.Element | null {
    const {
      actions,
      headerComponent
    } = this.props

    if (headerComponent || !actions) {
      return null
    }

    return (
      <DropdownMenu
        toggleComponent={this.dropdownMenuToggleComponent}
        sections={actions}
      />
    )
  }

  private dropdownMenuToggleComponent = (props: IDropdownMenuToggleComponentProps) => {
    return (
      <CommentActionMenuToggleButton
        type='button'
        onClick={props.toggleMenu}
        ref={props.toggleComponentRef}
        {...props.ariaProps}
      >
        <FontAwesomeIcon type='ellipsis-v' />
      </CommentActionMenuToggleButton>
    )
  }

  private get commentDate (): JSX.Element {
    const {
      dateComponent,
      pillComponent
    } = this.props

    return (
      <div className='comment-header-date'>
        {pillComponent}
        {dateComponent}
      </div>
    )
  }

  private get commentTitle (): JSX.Element {
    const {
      commentHeaderText,
      headerComponent
    } = this.props

    return (
      <div className='comment-header'>
        <span className='comment-header-person-name'>
          {commentHeaderText}
        </span>
        {headerComponent}
      </div>
    )
  }

  private get commentHeader (): JSX.Element {
    const {
      headerComponent
    } = this.props

    return (
      <div
        className={classNames(
          'comment-header-container',
          { 'with-status-update': !(!headerComponent) }
        )}
      >
        {this.commentTitle}

        {this.commentDate}

        {this.commentActions}
      </div>
    )
  }

  private get commentContent (): JSX.Element {
    const {
      commentBodyText
    } = this.props

    return (
      <div className='comment-content'>
        <FormattedText
          text={commentBodyText}
        />
      </div>
    )
  }
}
