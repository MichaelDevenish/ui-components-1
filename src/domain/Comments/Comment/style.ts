import React from 'react'
import styled, { StyledComponentClass } from 'styled-components'
import { getColor } from '../../../common/legacy'

const StyledComment = styled.div`

  display: flex;
  flex-flow: row;
  margin: 1.125rem 0;

  >.comment-badge-container {
    align-self: flex-start;
    padding: .5rem;
  }

  >.comment-container {
    align-self: stretch;
    background-color: ${getColor('neutral-light')};
    border-radius: .75rem;
    flex-grow: 1;
    padding: .85rem 1.2rem;

    >.comment-header-container {
      display: flex;
      flex-direction: row;
      flex-wrap: wrap;
      justify-content: flex-start;
      align-items: baseline;

      >.comment-header{
        >.comment-header-person-name {
          font-weight: 600;
        }
      }

      >.comment-header-date {
        color: ${getColor('badge-background-dark')};
        flex-grow: 1;
        font-size: 80%;
        margin-left: .4rem;

        >* {
          display: inline;
          margin: 0 .225rem;
        }
      }

      &.with-status-update {
        >.comment-header {
          flex-grow: 1;
        }
        >.comment-header-date {
          width: 100%;
          margin-left: -.15rem;
          margin-top: 1rem;
        }
      }
    }

    >.comment-content {
      margin-top: .5rem;
    }
  }
`

const CommentActionMenuToggleButton = styled.button`
  background-color: transparent;
  color: ${getColor('badge-background-dark')};
  cursor: pointer;
  padding: .2rem;
  transition: background-color .25s ease-out, color .25s ease-out;

  &:hover,
  &:active,
  &:focus {
    background-color: ${getColor('neutral-base')};
    color: ${getColor('main-text')};
    outline: none;
  }

  .fa {
    font-size: 1rem;
    line-height: 1;
    margin: 0;
    width: 1.3em;
  }
`

export {
  StyledComment,
  CommentActionMenuToggleButton
}
