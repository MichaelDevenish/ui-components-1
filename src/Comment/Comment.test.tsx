import { shallow } from 'enzyme'
import React from 'react'
import reactRenderer from 'react-test-renderer'

import { Comment } from './Comment'

describe('Comment', () => {
  it('Renders successfully a test message', () => {
    const component = reactRenderer.create(
      <Comment
        comment='Testing'
      />
    )
  
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})