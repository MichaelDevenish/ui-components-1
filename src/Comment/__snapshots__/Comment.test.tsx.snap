// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment Renders successfully a test message 1`] = `
<div>
  Testing
</div>
`;
